gobblin-core/src/test/java/gobblin/source/extractor/extract/AbstractSourceTest.java

```diff
@@ -0,0 +1,171 @@
+package gobblin.source.extractor.extract;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Properties;
+
+import gobblin.configuration.ConfigurationKeys;
+import gobblin.configuration.SourceState;
+import gobblin.configuration.WorkUnitState;
+import gobblin.source.extractor.Extractor;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.when;
+
+
+public class AbstractSourceTest {
+  
+  /**
+   * Test the never-retry policy.
+   */
+  @Test
+  public void testGetPreviousWorkUnitStatesNeverRetry() {
+    TestSource testSource = new TestSource();
+    SourceState mockedState = mock(SourceState.class);
```

Comment:
You don't need to mock a SourceState, you can just create an empty SourceState and populate properties into it.